### PR TITLE
Feature/branching support

### PR DESCRIPTION
**Pull request by JHabz**
Add branching support in Code Dx. In particular adds 2 fields, analysisBranch and analysisBranchParent which can be used either through the pipeline or freestyle project types

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/codedx/CodeDxPublisher.java [hunk 1]
```diff
@@ -261,10 +270,87 @@ public void perform(
...
+				String expandedBranchName = "";
+				String expandedBranchParent = "";
+
+				if (cdxVersion.compareTo(CodeDxVersion.MIN_FOR_BRANCHING) < 0) {
+					buildOutput.println("The connected Code Dx server is only version " + cdxVersion +
+							", which doesn't support branching (minimum supported version is " +
+							CodeDxVersion.MIN_FOR_BRANCHING + "). The analysis branch will not be set or sent.");
+				} else {
+					if (analysisBranch.length() == 0) {
+						analysisBranch = "DEFAULT";
```
Comment (baffles) [inline src/main/java/org/jenkinsci/plugins/codedx/CodeDxPublisher.java:282]: You shouldn't hard code this here; the default branch can be renamed or changed on the Code Dx side. If the intention here is just to enforce a default branch, I'd say just don't send anything to Code Dx in that case (if you never PUT to that branch endpoint, Code Dx will just use the default branch automatically).

File: src/main/java/org/jenkinsci/plugins/codedx/CodeDxPublisher.java [hunk 2]
```diff
@@ -261,10 +270,87 @@ public void perform(
...
+							boolean parentFound = false;
+							for (Branch branch : existingBranches) {
+								if (branch.getName().equals(expandedBranchParent)) {
+									buildOutput.println("Parent branch successfully found on Code Dx server");
+									parentFound = true;
+									break;
+								}
+							}
+							if (!parentFound) {
+								throw new IOException("The parent branch specified could not be found on the Code Dx server." +
```
Comment (baffles) [inline src/main/java/org/jenkinsci/plugins/codedx/CodeDxPublisher.java:333]: Why IOException? That doesn't seem like the right exception type

File: src/main/java/org/jenkinsci/plugins/codedx/CodeDxPublisher.java [hunk 3]
```diff
@@ -261,10 +270,87 @@ public void perform(
...
+										" Since the branch nor the parent branch could be found, Code Dx can not run");
+							}
+						} else {
+							throw new IOException("The branch you entered does not exist on the Code Dx server." +
```
Comment (baffles) [inline src/main/java/org/jenkinsci/plugins/codedx/CodeDxPublisher.java:337]: Same note, seems like the wrong type of exception to be throwing

File: src/main/java/org/jenkinsci/plugins/codedx/CodeDxPublisher.java [hunk 4]
```diff
@@ -261,10 +270,87 @@ public void perform(
...
+									" If this is intentional, then the parent branch must be specified");
+						}
+					} else {
+						if (analysisBranchParent.length() > 0) {
+							buildOutput.println("WARNING: The input branch already exists (" + expandedBranchName +
```
Comment (baffles) [inline src/main/java/org/jenkinsci/plugins/codedx/CodeDxPublisher.java:342]: This seems spammy; we wouldn't expect the user to figure out if the branch already exists in Code Dx, and this reads like a notification to the user that they're doing something wrong. It seems like the typical usage would be always specifying the parent, and then this plugin figuring out if it's relevant or not.

File: src/main/webapp/help-analysisBranch.html [hunk 5]
```diff
@@ -0,0 +1,16 @@
+<div>
+    <p>
+        This field lets you choose which Code Dx branch you would like the analysis to run on.
+        This does <b>not</b> need to match your repository or code branch name.
```
Comment (baffles) [inline src/main/webapp/help-analysisBranch.html:4]: It doesn't _need_ to, but it probably should match to prevent confusion

File: src/main/webapp/help-analysisBranch.html [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+        By default, Code Dx will run an analysis on the <code>DEFAULT</code> branch.
```
Comment (baffles) [inline src/main/webapp/help-analysisBranch.html:5]: Not sure if we want to mention this here? This technically isn't correct, as the default branch is configurable inside of Code Dx.

File: src/main/webapp/help-analysisBranchParent.html [hunk 7]
```diff
@@ -0,0 +1,13 @@
+<div>
+    <p>
+        This field is required only if the input Analysis Branch is a new branch. This field specifies which parent branch the new branch should be created from.
+    </p>
+    <p>
+        You can use build/environment variables to construct the parent for each analysis. In order to specify a new branch with parent <code>fix-parameters</code>:
+        <code>${GIT_BRANCH}</code>
+        would start an analysis for Code Dx branch with parent <code>fix-parameters</code>.
```
Comment (baffles) [inline src/main/webapp/help-analysisBranchParent.html:8]: How were you testing this? Were you able to formulate any sort of guidance on figuring this out using the git CLI from the links I had given?

We're going to need to know how to do this programatically for dogfooding, QA is going to need to know how to do this, and some basic advice/starting point for the user would be helpful.

**Review by baffles — CHANGES_REQUESTED** [review]

File: src/main/java/org/jenkinsci/plugins/codedx/CodeDxPublisher.java (hunk 2 above)
Comment (tylercamp) [reply src/main/java/org/jenkinsci/plugins/codedx/CodeDxPublisher.java:333]: Method signature on the calling method (`perform`) only allows `throw IOException, InterruptedException` since it's implementing a parent class, afaik don't have much of an option here

Comment (baffles) [reply src/main/java/org/jenkinsci/plugins/codedx/CodeDxPublisher.java:333]: It looks like Jenkins suggests [using https://javadoc.jenkins-ci.org/hudson/AbortException.html](https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/jenkins/tasks/SimpleBuildStep.java#L108).

File: src/main/java/org/jenkinsci/plugins/codedx/CodeDxPublisher.java [hunk 8]
```diff
@@ -278,9 +278,6 @@ public void perform(
 							", which doesn't support branching (minimum supported version is " +
 							CodeDxVersion.MIN_FOR_BRANCHING + "). The analysis branch will not be set or sent.");
 				} else {
-					if (analysisBranch.length() == 0) {
-						analysisBranch = "DEFAULT";
-					}
```
Comment (baffles) [inline src/main/java/org/jenkinsci/plugins/codedx/CodeDxPublisher.java]: Shouldn't this become `else if (analysisBranch.length() > 0)` so that we don't attempt to do anything with branch if there's no branch value? I would think the case of having no branch specified would match the case where Code Dx version isn't new enough to have branching.

File: src/main/java/com/secdec/codedx/api/client/CodeDxClient.java [hunk 9]
```diff
@@ -404,6 +404,7 @@ public StartAnalysisResponse startAnalysis(int projectId, String analysisBranch,
 		String prepId = createAnalysisPrep(projectId);
 
 		// If Code Dx supports branching
+		// If default branch is being used this may be skipped
 		if (analysisBranch.length() > 0) {
 			JsonObject branchBody = new JsonObject();
 			if (analysisBranchParent.length() > 0) {
```
Comment (baffles) [inline src/main/java/com/secdec/codedx/api/client/CodeDxClient.java:410]: It might not hurt to add null checks on these strings as well?

File: src/main/java/org/jenkinsci/plugins/codedx/CodeDxPublisher.java [hunk 10]
```diff
@@ -98,12 +99,16 @@ public CodeDxPublisher(
 			final String url,
 			final String key,
 			final String projectId,
-			final String analysisName
+			final String analysisName,
+			final String analysisBranch,
+			final String analysisBranchParent
```
Comment (baffles) [inline src/main/java/org/jenkinsci/plugins/codedx/CodeDxPublisher.java:104]: Please update the javadoc above for the added arguments. Wouldn't hurt to add `analysisName` to the javadoc as well since that got missed whenever it was added.

File: src/main/java/org/jenkinsci/plugins/codedx/CodeDxPublisher.java [hunk 11]
```diff
@@ -98,12 +99,16 @@ public CodeDxPublisher(
...
 	) {
 		this.projectId = projectId;
 		this.url = url;
 		this.key = key;
 		this.analysisName = analysisName.trim();
+		this.analysisBranch = analysisBranch.trim();
```
[the remark this answers is not in the document]
Comment (JHabz) [reply src/main/java/org/jenkinsci/plugins/codedx/CodeDxPublisher.java:110]: Yes, both arguments are optional (all parameters in jenkins are optional-we have to create errors ourselves to make them not optional) and they are set to the empty string in this file on lines 273-274. If branching isnt supported then these variables are never changed. Keep in mind the raw analysisName and analysisBranch are never passed to startAnalysis directly but the expanded variables do so I overwrite those instead

[the remark this answers is not in the document]
Comment (JHabz) [reply src/main/java/org/jenkinsci/plugins/codedx/CodeDxPublisher.java:110]: Also null is never passed, the empty string would be passed so we dont need to worry about null checking

[the remark this answers is not in the document]
Comment (JHabz) [reply src/main/java/org/jenkinsci/plugins/codedx/CodeDxPublisher.java:110]: Looks like null CAN get passed so I'll add null checking here

File: src/main/java/com/secdec/codedx/api/client/CodeDxClient.java (hunk 9 above)
Comment (JHabz) [reply src/main/java/com/secdec/codedx/api/client/CodeDxClient.java:410]: I now null check at the initial assignment from the user input fields so it would be redundant here

File: src/main/java/org/jenkinsci/plugins/codedx/CodeDxPublisher.java (hunk 8 above)
Comment (JHabz) [reply src/main/java/org/jenkinsci/plugins/codedx/CodeDxPublisher.java]: These two cases being handled the same way is OK because in either case we simply do not send anything to the `/branch` endpoint.

File: src/main/java/org/jenkinsci/plugins/codedx/CodeDxPublisher.java (hunk 10 above)
Comment (JHabz) [reply src/main/java/org/jenkinsci/plugins/codedx/CodeDxPublisher.java:104]: Will do, but this merge shouldn't be held up for this

Comment (baffles) [reply src/main/java/org/jenkinsci/plugins/codedx/CodeDxPublisher.java:104]: Easier than opening another PR, should be super quick (just 3 more lines up at line 92-96).

File: src/main/java/org/jenkinsci/plugins/codedx/CodeDxPublisher.java [hunk 12]
```diff
@@ -98,13 +99,27 @@ public CodeDxPublisher(
...
+			final String analysisName,
+			final String analysisBranch,
+			final String analysisBranchParent
 	) {
 		this.projectId = projectId;
 		this.url = url;
 		this.key = key;
 		this.analysisName = analysisName.trim();
 
+		if (analysisBranch != null) {
```
Comment (baffles) [inline src/main/java/org/jenkinsci/plugins/codedx/CodeDxPublisher.java:111]: If we can get nulls for these, does that mean we can also get a null for `analysisName` ? We probably should apply the same logic to that as well, since AFAIK it's an optional value.

File: src/main/java/com/secdec/codedx/api/client/CodeDxClient.java (hunk 9 above)
Comment (baffles) [reply src/main/java/com/secdec/codedx/api/client/CodeDxClient.java]: The suggestion was just more of a defensive thing in case some other piece of code ends up calling this. It'd be redundant, but not harmful, and lead to safer and slightly higher quality code.

File: src/main/java/org/jenkinsci/plugins/codedx/CodeDxPublisher.java (hunk 8 above)
Comment (baffles) [reply src/main/java/org/jenkinsci/plugins/codedx/CodeDxPublisher.java]: The downside is that you're doing all of this extra unnecessary work of querying for branches/checking for parent/etc. That is work that can easily be skipped by just adding a condition on this else path.

File: src/main/java/org/jenkinsci/plugins/codedx/CodeDxPublisher.java (hunk 12 above)
Comment (JHabz) [reply src/main/java/org/jenkinsci/plugins/codedx/CodeDxPublisher.java:116]: According to https://www.jenkins.io/doc/pipeline/steps/codedx/ it is not optional but it is clear from the code that it IS optional. This is another thing I will fix on that documentation... if I ever figure out how

tldr; it is optional but documentation is wrong

Comment (baffles) [reply src/main/java/org/jenkinsci/plugins/codedx/CodeDxPublisher.java:116]: I don't even know where that documentation is generated from. Maybe we should be doing something different for optional params or something (if that is autogenerated).

But yeah, code aside, knowing what it does and how it is handled on the Code Dx side shows that it's optional as well.

File: src/main/java/com/secdec/codedx/util/CodeDxVersion.java [hunk 13]
```diff
@@ -30,6 +30,9 @@
 	/** First version that supports the "analysis names" feature. */
 	public final static CodeDxVersion MIN_FOR_ANALYSIS_NAMES = fromString("2.4.0");
 
+	/** First version that supports branches */
+	public final static CodeDxVersion MIN_FOR_BRANCHING = fromString("2022.1.0");
```
Comment (baffles) [inline src/main/java/com/secdec/codedx/util/CodeDxVersion.java:34]: Also, Tyler pointed this out on Slack, but this value is wrong. It should be 2022.4.0. The version number for the accepted/branching-support branch needs to be updated to 2022.4.0 (which I will work on now).

File: src/main/webapp/help-analysisBranchParent.html [hunk 14]
```diff
@@ -0,0 +1,13 @@
...
+    </p>
+    <p>
+        If the entered child branch already exists, then this field will be ignored but will log a warning.
```
Comment (baffles) [inline src/main/webapp/help-analysisBranchParent.html:11]: Outdated doc if that warning was removed?

**Review by baffles — CHANGES_REQUESTED** [review]
Seems like a lot of my original comments remain unaddressed.

* Looks like we should be using AbortException and not IOException everywhere
* Easy tweak to update javadoc
* Presumably easy tweak to not do extra branching lookups unnecessarily
* Applying null filtering to analysis name for consistency

File: src/main/java/org/jenkinsci/plugins/codedx/CodeDxPublisher.java (hunk 12 above)
Comment (tylercamp) [reply src/main/java/org/jenkinsci/plugins/codedx/CodeDxPublisher.java:116]: (Mentioned this to Jared on Slack, but those docs are auto-generated by Jenkins via reflection)

Comment (tylercamp) [reply src/main/java/org/jenkinsci/plugins/codedx/CodeDxPublisher.java:116]: Having a parameter in the constructor apparently marks it as "required" to Jenkins - https://www.jenkins.io/doc/developer/plugin-development/pipeline-integration/#constructor-vs-setters

It's not apparent whether `null` is a valid value for a required param. @JHabz Can you test locally and confirm for a pipeline project? Freestyle should fill with empty string if left unassigned, so I don't think that needs to be tested.

Comment (tylercamp) [reply src/main/java/org/jenkinsci/plugins/codedx/CodeDxPublisher.java:116]: (For this PR I think the param should be left as-is, there's other Jenkins plugin cleanup tasks and changing that param can be an extra task in the list.)

Comment (baffles) [reply src/main/java/org/jenkinsci/plugins/codedx/CodeDxPublisher.java:116]: I think that we should consistently handle this sort of thing, at least in the context of a single unit. So either the new code doesn't have a null check, or the old code is updated with a null check. No reason to kick tech debt cans down the road when it's a change that takes 20 seconds to put in place while the code is being touched in this area.

Comment (baffles) [reply src/main/java/org/jenkinsci/plugins/codedx/CodeDxPublisher.java:116]: I'd say at the very least (and in general, if a change is being skipped due to concern about it breaking something or having testing impact), there should be a TODO comment acknowledging that this change should probably be applied, that way someone isn't scratching their head later wondering why out of these 3 seemingly optional parameters, only two have a null check.

But an extra null check doesn't seem particularly risky to add, and it would take longer to write such a TODO comment, so I'd just make the tweak now.

Comment (JHabz) [reply src/main/java/org/jenkinsci/plugins/codedx/CodeDxPublisher.java:116]: I noticed that I would need to create a setter with `@DataBoundSetter` to make both my 2 new parameters as well as `analysisName` optional. I agree with Bobby in that fixing it is probably is the best way to go about it. We don't want people thinking certain params are required when they are not.

I did test the null thing locally and yes in freestyle its the empty string but if you don't add the param in pipeline it is null so the check is required

**Review by baffles — APPROVED** [review]
